File: galaxy-web-app/galaxy-oj/src/main/java/io/charlie/web/oj/modular/data/library/entity/DataLibrary.java
package io.charlie.web.oj.modular.data.library.entity;

import com.baomidou.mybatisplus.annotation.TableField;
import com.baomidou.mybatisplus.annotation.TableName;
import com.baomidou.mybatisplus.extension.handlers.JacksonTypeHandler;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import io.charlie.galaxy.config.timestamp.DateToTimestampSerializer;
import io.charlie.galaxy.config.timestamp.TimestampToDateDeserializer;
import io.charlie.galaxy.pojo.CommonEntity;

import java.io.Serial;
import java.util.Date;
import java.util.List;

import io.charlie.web.oj.modular.data.problem.entity.DataProblem;
import io.charlie.web.oj.modular.data.set.entity.DataSet;
import io.charlie.web.oj.modular.sys.user.entity.SysUser;
import io.swagger.v3.oas.annotations.media.Schema;
import lombok.Data;
import lombok.EqualsAndHashCode;
import org.dromara.core.trans.anno.Trans;
import org.dromara.core.trans.constant.TransType;

/**
 * @author Charlie Zhang
 * @version v1.0
 * @date 2025-09-20
 * @description 提交样本库
 */
@EqualsAndHashCode(callSuper = true)
@Data
@TableName(value = "data_library", autoResultMap = true)
@Schema(name = "DataLibrary", description = "提交样本库")
public class DataLibrary extends CommonEntity {
    @Serial
    @TableField(exist = false)
    private static final long serialVersionUID = 1L;

    @Schema(description = "主键")
    private String id;

    @Schema(description = "用户ID")
    @Trans(type = TransType.SIMPLE, target = SysUser.class, fields = {"nickname", "avatar"}, refs = {"userIdName", "userAvatar"})
    private String userId;

    @Schema(description = "用户名称")
    @TableField(exist = false)
    private String userIdName;

    @Schema(description = "用户头像")
    @TableField(exist = false)
    private String userAvatar;

    @Schema(description = "题集ID")
    @Trans(type = TransType.SIMPLE, target = DataSet.class, fields = "title", ref = "setIdName")
    private String setId;

    @Schema(description = "题集名称")
    @TableField(exist = false)
    private String setIdName;

    @Schema(description = "是否是题集提交")
    @Trans(type = TransType.DICTIONARY, key = "YES_NO")
    private Boolean isSet;

    @Schema(description = "题目ID")
    @Trans(type = TransType.SIMPLE, target = DataProblem.class, fields = "title", ref = "problemIdName")
    private String problemId;

    @Schema(description = "题目名称")
    @TableField(exist = false)
    private String problemIdName;

    @Schema(description = "提交ID")
    private String submitId;

    @Schema(description = "提交时间")
    @JsonSerialize(using = DateToTimestampSerializer.class)
    @JsonDeserialize(using = TimestampToDateDeserializer.class)
    private Date submitTime;

    @Schema(description = "编程语言")
    @Trans(type = TransType.DICTIONARY, key = "ALLOW_LANGUAGE", ref = "languageName")
    private String language;

    @Schema(description = "语言名称")
    @TableField(exist = false)
    private String languageName;

    @Schema(description = "源代码")
    private String code;

    @Schema(description = "源代码长度")
    private Integer codeLength;

    @TableField(typeHandler = JacksonTypeHandler.class)
    private List<Integer> codeToken;

    @TableField(typeHandler = JacksonTypeHandler.class)
    private List<String> codeTokenName;

    @TableField(typeHandler = JacksonTypeHandler.class)
    private List<String> codeTokenTexts;

    @Schema(description = "访问次数")
    private Integer accessCount;
}

File: similarity-service/internal/database/model/data_library.go
package model

import (
	"time"
)

// 提交样本库
type DataLibrary struct {
	ID             string     `gorm:"column:id;primaryKey;type:varchar(32)"`
	UserID         string     `gorm:"column:user_id;type:varchar(32)"`
	SetID          string     `gorm:"column:set_id;type:varchar(32)"`
	IsSet          bool       `gorm:"column:is_set;default:false"`
	ProblemID      string     `gorm:"column:problem_id;type:varchar(32)"`
	SubmitID       string     `gorm:"column:submit_id;type:varchar(32)"`
	SubmitTime     *time.Time `gorm:"column:submit_time"`
	Language       string     `gorm:"column:language;type:varchar(64)"`
	Code           string     `gorm:"column:code;type:longtext"`
	CodeToken      string     `gorm:"column:code_token;type:json"` // 实际使用时可定义为具体的结构体类型
	CodeTokenName  string     `gorm:"column:code_token_name;type:json"`
	CodeTokenTexts string     `gorm:"column:code_token_texts;type:json"`
	CodeLength     int        `gorm:"column:code_length;default:0"`
	AccessCount    int        `gorm:"column:access_count;default:0"`
	Deleted        bool       `gorm:"column:deleted;default:false"`
	CreateTime     *time.Time `gorm:"column:create_time"`
	CreateUser     string     `gorm:"column:create_user;type:varchar(32)"`
	UpdateTime     *time.Time `gorm:"column:update_time"`
	UpdateUser     string     `gorm:"column:update_user;type:varchar(32)"`
}

func (DataLibrary) TableName() string {
	return "data_library"
}
